chrome/android/java/src/org/chromium/chrome/browser/adblock/AdblockInitializer.java

```diff
@@ -0,0 +1,140 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Orca.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.chromium.chrome.browser.adblock;
+
+import android.content.Context;
+import android.support.annotation.UiThread;
+
+import org.adblockplus.libadblockplus.android.AdblockEngine;
+import org.adblockplus.libadblockplus.android.SingleInstanceEngineProvider;
+import org.adblockplus.libadblockplus.android.settings.AdblockHelper;
+import org.chromium.base.ContextUtils;
+import org.chromium.base.ThreadUtils;
+import org.chromium.base.library_loader.ProcessInitException;
+import org.chromium.chrome.browser.init.BrowserParts;
+import org.chromium.chrome.browser.init.ChromeBrowserInitializer;
+import org.chromium.chrome.browser.init.EmptyBrowserParts;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public final class AdblockInitializer implements SingleInstanceEngineProvider.EngineCreatedListener {
+
+    private final InitializationState mUnitializedState = new UnitializedState();
+    private final InitializationState mInitializedState = new InitializedState();
+    private InitializationState mInitializationState = mUnitializedState;
+    private final Set<OnAdblockInitListener> mInitListeners = new HashSet<>();
+
+    private AdblockInitializer() {
+    }
+
+    // Thread-safe lazy initialization
+    private static class LazyLoader {
+        private static final AdblockInitializer sInstance = new AdblockInitializer();
+    }
+
+    public static AdblockInitializer getInstance() {
+        return AdblockInitializer.LazyLoader.sInstance;
+    }
+
+    @Override
+    public void onAdblockEngineCreated(final AdblockEngine engine) {
+        ThreadUtils.postOnUiThread(() -> beInitialized(engine));
+    }
+
+    @UiThread
+    public void init(final ChromeBrowserInitializer initializer) {
+        mInitializationState.init(initializer);
+    }
+
+    public void addAdblockInitListener(final OnAdblockInitListener listener) {
+        mInitializationState.addAdblockInitListener(listener);
+    }
+
+    private void initAdblock() {
+        final Context context = ContextUtils.getApplicationContext();
+        final String basePath = context.getDir(
+                AdblockEngine.BASE_PATH_DIRECTORY,
+                Context.MODE_PRIVATE)
+                .getAbsolutePath();
+
+        final long ptr = AdblockBridge.getInstance().getIsolateProviderNativePtr();
+        AdblockHelper.get()
+                .init(context, basePath, true, AdblockHelper.PREFERENCE_NAME)
+                .addEngineCreatedListener(this)
+                .useV8IsolateProvider(ptr);
+
+        AdblockHelper.get().getProvider().retain(true);
+    }
+
+    @UiThread
+    private void beInitialized(final AdblockEngine engine) {
```

Comment:
i'd vote for better naming. Is it called when initialized? What about 'onInitialized' or smth similar then?

Reply:
This is called when the creation of the Engine is complete, so it would change its state to Initialized. `onInitialized` gives the idea that the initialization has been complete already. Maybe something else, like `completeInitialization` or  `finishInitialization`?

Reply:
yes, `onInitialized` sounds much better for me